lib/at_host/at_host.c

```diff
@@ -53,28 +53,50 @@ enum select_uart {
 
 static enum term_modes term_mode;
 static struct device *uart_dev;
-static int at_socket_fd = INVALID_DESCRIPTOR;
-static struct pollfd fds[1];
-static int nfds;
 static u8_t at_buf[AT_MAX_CMD_LEN];
 static size_t at_buf_len;
-static struct k_work at_cmd_send_work;
-static struct k_thread socket_thread;
-static K_THREAD_STACK_DEFINE(socket_thread_stack, THREAD_STACK_SIZE);
-
-
+static struct k_work cmd_send_work;
 static const char termination[3] = { '\0', '\r', '\n' };
 
-static void at_cmd_send(struct k_work *work)
+static inline void write_uart_string(char *str, size_t len)
 {
-	int bytes_sent;
+	for (size_t i = 0; i < len; i++) {
+		uart_poll_out(uart_dev, str[i]);
+	}
+}
+
+static void cmd_send(struct k_work *work)
+{
+	int    err;
+	size_t chars;
+	char   str[15];
 
 	ARG_UNUSED(work);
 
-	bytes_sent = send(at_socket_fd, at_buf, at_buf_len, 0);
+	err = at_cmd_write(at_buf);
+	if (err != AT_CMD_OK) {
+		LOG_ERR("Could not send AT command to modem: %d", err);
+	}
 
-	if (bytes_sent <= 0) {
-		LOG_ERR("Could not send AT command to modem: %d", bytes_sent);
+	switch (err) {
+	case AT_CMD_OK:
+		chars = sprintf(str, "OK\r\n");
+		write_uart_string(str, chars);
+		break;
+	case AT_CMD_ERROR:
+		chars = sprintf(str, "ERROR\r\n");
+		write_uart_string(str, chars);
+		break;
+	case AT_CMD_CMS:
+		chars = sprintf(str, "+CMS: %d\r\n", at_cmd_get_last_error());
```

Comment:
Could we not get the original response String from the Modem instead of creating it?

Reply:
In most use-cases the application want to act upon the return code of an at command call, and then process returned data if any. So I kind of view the at_host library as an exception to what the common use-cases for this driver will be. 

If I return the whole buffer it kind of defy the purpose of the interface driver because that is what most applications will want to do.

Reply:
Yes, it is good to have the return code here.
You need to recreate the response manually here instead of printing the response from the Modem. Can't we read the Modem response directly?